### PR TITLE
Hourly FFMC

### LINKED ISSUE #3517: SFMS: Hourly FFMC
**As a** Fire Behaviour Specialist
**I need** to project the FFMC hourly in the future
**So That** I can specify what times of day Fire Behaviour Advisories apply

**Acceptance Criteria**
- [ ] Given we have a SFMS grid output for the previous hour's observed Hourly Fine Fuel Moisture Code (HFFMC), when a new weather forecast arrives, then we can use this as an input to iteratively calculate HFFMC intro the future.
- [ ] The HFFMC calculations should only be applied when a weather forecast is issued at 00Z and 12Z.
- [ ] For the 00Z forecast interval, grab the 4 PM Pacific observed HFFMC as the starting condition and calculate forecasted HFMMC values at least 43 hours into the future from 00Z (or 5 PM Pacific today)
- [ ] For the 12Z forecast interval, we will grab the 4 AM Pacific observed HFFMC as the starting condition and calculate forecasted HFMMC values at least 31 hours into the future from 12Z (or 5 AM Pacific tomorrow). Overwrite/replace all previous HFFMC forecast grids since actuals and a new forecast are available.

**Additional Context**
- HFFMC Raster calculations require hour by hour grid inputs for: temperature, relative humidity, wind speed, and precipitation. A starting HFFMC grid is also required for the first calculation: https://cwfis.cfs.nrcan.gc.ca/background/summary/fwi
- The Regional weather forecast has data 48 hours into the future. If data storage is not an issue, then run the HFFMC calculations 48 hours into the future. The 43 and 31-hour acceptance criteria for 00Z and 12Z forecast intervals are the minimum requirements to ensure  there are HFFMC data for all of DAY 1 and the overnight/early morning period into DAY 2.
> - 00Z Forecast: Day 0 at 5PM + 43 hours = DAY 2 at 12 Noon
> - 12Z Forecast: Day 1 at 5AM + 31 hours = DAY 2 at 12 Noon
- Archiving & Retention Needs: everything will need to be stored + saved; Leapfrogging Policy
- We may need to produce our own actuals, given that SFMS is not producing outputs currently 

**Definition of Done**
https://github.com/bcgov/wps/wiki/Definition-of-Done

### PR DESCRIPTION
**Pull request by dgboss**
To test locally, set `MAX_MODEL_RUN_HOUR` to 12 or so in `rdps_sfms.py` and run/debug the sfms raster processor job option. You might want to delete `sfms/calculated/hourlies` from dev S3 storage as well to confirm the new hFFMC rasters are being generated.

Fixed: ~~There is a known bug with warping rH rasters where it is possible to have rH values greater than 100. This breaks hFFMC calculations because the underlying cffdrs ffmc calculation has a range check on rH values to make sure they are between 0 and 100. I'll address this issue in another PR.~~

Calculate and store hourly FFMC rasters.

The general process is:
1. cron job kicks off the job and we use current UTC time as start time
2. Create HourlyFFMCProcessor with the start time and begin processing
3. Use job start time to determine most recent RDPS model run start time (date and 00z or 12z)
4. Use most recent RDPS model run start time to determine most recent hFFMC key to use as source which is always one hour before the RDPS start time (04 or 16 PDT)
5. Start calculating hFFMC from model run hour 0 through to 47. Save the calculated hFFMCs to S3. Most recently calculated hFFMC is used as input to the next hour's hFFMC calculation.
6. hFFMC rasters are saved to S3 with PDT based keys.

closes #3517 

# Test Links:
[Landing Page](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/morecast)
[Percentile Calculator](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/percentile-calculator)
[C-Haines](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-4153-e1e498-dev.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: api/app/sfms/fwi_processor.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def calculate_ffmc(previous_ffmc_ds: WPSDataset, temp_ds: WPSDataset, rh_ds: WPS
     wind_speed_array, _ = wind_speed_ds.replace_nodata_with(0)
 
     start = perf_counter()
-    ffmc_values = vectorized_ffmc(previous_ffmc_array, temp_array, rh_array, precip_array, wind_speed_array)
+    ffmc_values = vectorized_ffmc(previous_ffmc_array, temp_array, rh_array, wind_speed_array, precip_array)
```
Comment (brettedw) [inline api/app/sfms/fwi_processor.py:73]: Nice catch!

File: api/app/sfms/hourly_ffmc_processor.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        # 1 - Determine starting hFFMC (4am or 4pm) from SFMS and get key, confirm exists, if not, exit
+        # 2 - Determine what would be last key of run and check if exists, if exists, exit
+        # 3 - Get all weather variable keys and check if last one exists, if not, exit
+        # 4 - Use seed hFFMC plus:
+        #       - rh, temp and wind speed from RDPS model run hour n = 000
+        #       - computed precip at n = 0Z or 12Z
+        # 5 - Use newly calculated hFFMC to calculate next hFFMC using:
+        #       -  rh, temp and wind speed from RDPS model run hour n + 1
+        #       - computed precip at n + 1
+        # hFFMC files from SFMS use PST datetimes
```
Comment (brettedw) [inline api/app/sfms/hourly_ffmc_processor.py:44]: I think our SFMS uploads use whatever local time is in Vancouver's time zone, so during the fire season these would be uploaded in PDT (see app/utils/sfms.py). I think your still grabbing the correct input data though, if my mental timezone converting is correct

Comment (dgboss) [reply api/app/sfms/hourly_ffmc_processor.py]: Fixed in https://github.com/bcgov/wps/pull/4153/commits/0e7bda3053128b3150a09055b2964353eb2a82a9

**Review by conbrad — APPROVED** [review]
Nice work!

File: api/app/sfms/hourly_ffmc_processor.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    async def process(self, s3_client: S3Client, input_dataset_context: MultiDatasetContext, hours_to_process: int = MAX_MODEL_RUN_HOUR):
+        set_s3_gdal_config()
+
+        # hFFMC general process
+        # 1. cron job kicks off the job and we use current UTC time as start time
+        # 2. Create HourlyFFMCProcessor with the start time and begin processing
+        # 3. Use job start time to determine most recent RDPS model run start time (date and 00z or 12z)
+        # 4. Use most recent RDPS model run start time to determine most recent hFFMC key to use as source which is always one hour before the RDPS start time (04 or 16 PDT)
+        # 5. Start calculating hFFMC from model run hour 0 through to 47. Save the calculated hFFMCs to S3. Most recently calculated hFFMC is used as input to the next hour's hFFMC calculation.
+        # 6. hFFMC rasters are saved to S3 with PDT based keys.
```
[the remark this answers is not in the document]
Comment (dgboss) [reply api/app/sfms/hourly_ffmc_processor.py:41]: The hourly SFMS rasters we receive are named with PDT times so I followed the existing pattern. I'm totally open to naming via UTC timestamp instead. My brain definitely prefers to work in UTC as it never changes.

[the remark this answers is not in the document]
Comment (dgboss) [reply api/app/sfms/hourly_ffmc_processor.py:41]: Oh, and our other calcualted SFMS S3 rasters/keys are daily in nature, so need to consider UTC vs PDT in the naming convention.

File: api/app/sfms/daily_fwi_processor.py [hunk 4]
```diff
@@ -105,7 +106,7 @@ async def process(
                     )
 
                     # Create and store FFMC dataset
-                    ffmc_values, ffmc_no_data_value = calculate_ffmc(ffmc_ds, warped_temp_ds, warped_rh_ds, warped_wind_speed_ds, warped_precip_ds)
+                    ffmc_values, ffmc_no_data_value = calculate_ffmc(ffmc_ds, warped_temp_ds, warped_rh_ds, warped_precip_ds, warped_wind_speed_ds)
```
[the remark this answers is not in the document]
Comment (dgboss) [reply api/app/sfms/daily_fwi_processor.py:109]: We could add some sort of name of id property to `WPSDataset` objects and use this to assert that we're getting parameters in the right order?

File: api/app/sfms/hourly_ffmc_processor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (conbrad) [reply api/app/sfms/hourly_ffmc_processor.py:41]: Right SFMS gives us PDT rasters. Either way works but I think adding the timezone to the naming convention would be helpful.

File: api/app/sfms/daily_fwi_processor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (conbrad) [reply api/app/sfms/daily_fwi_processor.py:109]: Yeah I think that'd be worth it since we're passing so many `WPSDataset`'s around

**Pull request merged**